### DISCUSSION TITLE
How to parallel run multiple background task?

### DISCUSSION THREAD

**Discussion by antx-code** [fastapi/fastapi#6671] [Questions] [answered]
### Code snippet:

from fastapi import APIRouter, Depends, BackgroundTasks
from pydantic import BaseModel
from loguru import logger

@logger.catch(level='ERROR')
@router.post('/create_task')
async def create_task(query_parm: CtaskModel, background_task: BackgroundTasks):
   """

  some base code.

   """
    background_task.add_task(ctask.run, target=[targets['target'], targets['website'].strip()], target_type=task_type, target_url=target_url)
    background_task.add_task(vuln_verifier.dia, targets=[target_url], filter=False)
    return msg(status='success', data=f'已成功创建{task_type}-website任务')

### Description

I want to run ctask function and vuln_verifier function background task at the same time, but it run one by one. 
How to parallel run background task A and background B ? The Task A(funciton ctask for me) and the task B(function vuln_verifier for me) are python package script I wrote.

### Environment

* OS: manjaro 18.0.4 
* FastAPI version: 0.63.0
* Python version: 3.8
* Uvicorn version: 0.11.8

**Comment by gyKa** [comment]
Hi,

I didn't find easy solution, but based on documentation you can use [Celery and message broker](https://fastapi.tiangolo.com/tutorial/background-tasks/?h=backgr#caveat).

**Comment by antx-code** [comment]
@gyKa Thank you for your answer. I have two methods now.
M1: use Celery and message broker
M2: develop Task A and Task B as gRPC service

**Comment by JeongHyeongKim** [comment]
https://apscheduler.readthedocs.io/en/stable/index.html
how about this
i research background task for batch service and i found the dependency above.

**Comment by antx-code** [comment]
> https://apscheduler.readthedocs.io/en/stable/index.html
> how about this
> i research background task for batch service and i found the dependency above.

But it seems that just can scheduler tasks periodically, can not parallel excute different task in one time.

**Comment by YuriiMotov** [comment, marked as answer]
If the functions you want to run are `async`, you can use the approach described in this comment: https://github.com/fastapi/fastapi/discussions/10682#discussioncomment-13191413

For `sync` functions you can modify that code to run functions in thread pool and gather and await results
